File: letterCombinations.java
import java.util.*;
class Solution{
	Map<Character,String> map = Map.of('2',"abc",'3',"def",'4',"ghi",'5',"jkl",'6',"mno",
		'7',"pqr",'8',"stuv",'9',"wxyz");

	public void backtrack(List<String> ans, StringBuilder comb, String digits, int index){
		if(comb.length()==digits.length()){
			ans.add(comb.toString());
			return;
		}
		String lettersPossible=map.get(digits.charAt(index));
		for(char c: lettersPossible.toCharArray()){
			comb.append(c);
			backtrack(ans,comb,digits,index+1);
			comb.deleteCharAt(comb.length()-1);
		}

	}

	public List<String> letterCombinations(String digits){
		if(digits.length()==0) return new ArrayList<>();
		List<String> ans= new ArrayList<>();
		StringBuilder comb= new StringBuilder();
		backtrack(ans,comb,digits,0);
		return ans;
	}
}

class letterCombinations{
	public static void main(String[] args) {
		String digits="23";
		Solution s= new Solution();
		System.out.println(s.letterCombinations(digits));
	}
}

File: letterCombinations.cpp
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    unordered_map<char, string> map={{'2',"abc"},{'3',"def"},{'4',"ghi"},
    {'5',"jkl"},{'6',"mno",}, {'7',"pqrs"},{'8',"tuv"},{'9',"wxyz"}};
    
    vector<string> ans;

    vector<string> letterCombinations(string digits) {
        if (digits.size()==0) return ans;
        string path="";
        backtrack(0, path, digits);
        return ans;
        
    }

    void backtrack(int index, string path, string digits){
        if(path.size()==digits.size()){
            ans.push_back(path);
            return;
        }

        for(auto& c: map[digits[index]]){
            path+=c; 
            backtrack(index+1, path, digits);
            path.erase(path.size()-1);
        }
    }
};

int main(){
    Solution s;
    for(auto& s: s.letterCombinations("23")){
        cout << s << ", ";
    }
    cout << endl;
    return 0;

}